File: c++/log_test.cc
#include <gtest/gtest.h>
#include <thread>

#include "log.h"
#include "memkvstore.h"
#include "test_util.h"

using multipaxos::Command;
using multipaxos::CommandType;
using multipaxos::CommandType::DEL;
using multipaxos::CommandType::GET;
using multipaxos::CommandType::PUT;

using multipaxos::Instance;
using multipaxos::InstanceState;
using multipaxos::InstanceState::COMMITTED;
using multipaxos::InstanceState::EXECUTED;
using multipaxos::InstanceState::INPROGRESS;

class LogTest : public testing::Test {
 public:
  LogTest() : log_(std::make_unique<kvstore::MemKVStore>()) {}

 protected:
  Log log_;
};

using LogDeathTest = LogTest;

TEST_F(LogTest, Constructor) {
  EXPECT_EQ(log_.LastExecuted(), 0);
  EXPECT_EQ(log_.GlobalLastExecuted(), 0);
  EXPECT_FALSE(log_.IsExecutable());
  EXPECT_EQ(nullptr, log_.at(0));
  EXPECT_EQ(nullptr, log_.at(-1));
  EXPECT_EQ(nullptr, log_.at(3));
}

TEST_F(LogTest, Insert) {
  std::unordered_map<int64_t, multipaxos::Instance> log;
  auto index = 1;
  auto ballot = 1;
  EXPECT_TRUE(Insert(&log, MakeInstance(ballot, index, PUT)));
  EXPECT_EQ(PUT, log[index].command().type());
  EXPECT_FALSE(Insert(&log, MakeInstance(ballot, index, PUT)));
}

TEST_F(LogTest, InsertUpdateInProgress) {
  std::unordered_map<int64_t, multipaxos::Instance> log;
  auto index = 1;
  auto ballot = 1;
  EXPECT_TRUE(Insert(&log, MakeInstance(ballot, index, PUT)));
  EXPECT_EQ(PUT, log[index].command().type());
  EXPECT_FALSE(Insert(&log, MakeInstance(ballot + 1, index, DEL)));
  EXPECT_EQ(DEL, log[index].command().type());
}

TEST_F(LogTest, InsertUpdateCommitted) {
  std::unordered_map<int64_t, multipaxos::Instance> log;
  auto index = 1;
  auto ballot = 1;
  EXPECT_TRUE(Insert(&log, MakeInstance(ballot, index, COMMITTED, PUT)));
  EXPECT_FALSE(Insert(&log, MakeInstance(ballot, index, INPROGRESS, PUT)));
}

TEST_F(LogTest, InsertStale) {
  std::unordered_map<int64_t, multipaxos::Instance> log;
  auto index = 1;
  auto ballot = 1;
  EXPECT_TRUE(Insert(&log, MakeInstance(ballot, index, PUT)));
  EXPECT_EQ(PUT, log[index].command().type());
  EXPECT_FALSE(Insert(&log, MakeInstance(ballot - 1, index, DEL)));
  EXPECT_EQ(PUT, log[index].command().type());
}

TEST_F(LogDeathTest, InsertCase2Committed) {
  auto index = 1;
  auto inst1 = MakeInstance(0, index, COMMITTED, PUT);
  auto inst2 = MakeInstance(0, index, INPROGRESS, DEL);
  std::unordered_map<int64_t, multipaxos::Instance> log;
  Insert(&log, std::move(inst1));
  EXPECT_DEATH(Insert(&log, std::move(inst2)), "Insert case2");
}

TEST_F(LogDeathTest, InsertCase2Executed) {
  auto index = 1;
  auto inst1 = MakeInstance(0, index, EXECUTED, PUT);
  auto inst2 = MakeInstance(0, index, INPROGRESS, DEL);
  std::unordered_map<int64_t, multipaxos::Instance> log;
  Insert(&log, std::move(inst1));
  EXPECT_DEATH(Insert(&log, std::move(inst2)), "Insert case2");
}

TEST_F(LogDeathTest, InsertCase3) {
  auto index = 1;
  auto inst1 = MakeInstance(0, index, INPROGRESS, PUT);
  auto inst2 = MakeInstance(0, index, INPROGRESS, DEL);
  std::unordered_map<int64_t, multipaxos::Instance> log;
  Insert(&log, std::move(inst1));
  EXPECT_DEATH(Insert(&log, std::move(inst2)), "Insert case3");
}

TEST_F(LogTest, Append) {
  log_.Append(MakeInstance(0, log_.AdvanceLastIndex()));
  log_.Append(MakeInstance(0, log_.AdvanceLastIndex()));
  EXPECT_EQ(1, log_.at(1)->index());
  EXPECT_EQ(2, log_.at(2)->index());
}

TEST_F(LogTest, AppendWithGap) {
  auto index = 42;
  log_.Append(MakeInstance(0, index));
  EXPECT_EQ(index, log_.at(index)->index());
  EXPECT_EQ(index + 1, log_.AdvanceLastIndex());
}

TEST_F(LogTest, AppendFillGaps) {
  auto index = 42;
  log_.Append(MakeInstance(0, index));
  log_.Append(MakeInstance(0, index - 10));
  EXPECT_EQ(index + 1, log_.AdvanceLastIndex());
}

TEST_F(LogTest, AppendHighBallotOverride) {
  auto index = 1, lo_ballot = 0, hi_ballot = 1;
  log_.Append(MakeInstance(lo_ballot, index, PUT));
  log_.Append(MakeInstance(hi_ballot, index, DEL));
  EXPECT_EQ(DEL, log_.at(index)->command().type());
}

TEST_F(LogTest, AppendLowBallotNoEffect) {
  auto index = 1, lo_ballot = 0, hi_ballot = 1;
  log_.Append(MakeInstance(hi_ballot, index, PUT));
  log_.Append(MakeInstance(lo_ballot, index, DEL));
  EXPECT_EQ(PUT, log_.at(index)->command().type());
}

TEST_F(LogTest, Commit) {
  auto index1 = 1;
  log_.Append(MakeInstance(0, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(0, index2));

  EXPECT_TRUE(IsInProgress(*log_.at(index1)));
  EXPECT_TRUE(IsInProgress(*log_.at(index2)));
  EXPECT_FALSE(log_.IsExecutable());

  log_.Commit(index2);

  EXPECT_TRUE(IsInProgress(*log_.at(index1)));
  EXPECT_TRUE(IsCommitted(*log_.at(index2)));
  EXPECT_FALSE(log_.IsExecutable());

  log_.Commit(index1);

  EXPECT_TRUE(IsCommitted(*log_.at(index1)));
  EXPECT_TRUE(IsCommitted(*log_.at(index2)));
  EXPECT_TRUE(log_.IsExecutable());
}

TEST_F(LogTest, CommitBeforeAppend) {
  auto index = 1;
  std::thread commit_thread([this, index] { log_.Commit(index); });
  std::this_thread::yield();
  log_.Append(MakeInstance(0, log_.AdvanceLastIndex()));
  commit_thread.join();
  EXPECT_TRUE(IsCommitted(*log_.at(index)));
}

TEST_F(LogTest, AppendCommitExecute) {
  std::thread execute_thread([this] { log_.Execute(); });

  auto index = 1;
  log_.Append(MakeInstance(0, index));
  log_.Commit(index);
  execute_thread.join();

  EXPECT_TRUE(IsExecuted(*log_.at(index)));
  EXPECT_EQ(index, log_.LastExecuted());
}

TEST_F(LogTest, AppendCommitExecuteOutOfOrder) {
  std::thread execute_thread([this] {
    log_.Execute();
    log_.Execute();
    log_.Execute();
  });

  auto index1 = 1;
  log_.Append(MakeInstance(0, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(0, index2));
  auto index3 = 3;
  log_.Append(MakeInstance(0, index3));

  log_.Commit(index3);
  log_.Commit(index2);
  log_.Commit(index1);

  execute_thread.join();

  EXPECT_TRUE(IsExecuted(*log_.at(index1)));
  EXPECT_TRUE(IsExecuted(*log_.at(index2)));
  EXPECT_TRUE(IsExecuted(*log_.at(index3)));
  EXPECT_EQ(index3, log_.LastExecuted());
}

TEST_F(LogTest, CommitUntil) {
  auto ballot = 0;
  auto index1 = 1;
  log_.Append(MakeInstance(ballot, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(ballot, index2));
  auto index3 = 3;
  log_.Append(MakeInstance(ballot, index3));

  log_.CommitUntil(index2, ballot);

  EXPECT_TRUE(IsCommitted(*log_.at(index1)));
  EXPECT_TRUE(IsCommitted(*log_.at(index2)));
  EXPECT_FALSE(IsCommitted(*log_.at(index3)));
  EXPECT_TRUE(log_.IsExecutable());
}

TEST_F(LogTest, CommitUntilHigherBallot) {
  auto ballot = 0;
  auto index1 = 1;
  log_.Append(MakeInstance(ballot, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(ballot, index2));
  auto index3 = 3;
  log_.Append(MakeInstance(ballot, index3));

  log_.CommitUntil(index3, ballot + 1);

  EXPECT_FALSE(IsCommitted(*log_.at(index1)));
  EXPECT_FALSE(IsCommitted(*log_.at(index2)));
  EXPECT_FALSE(IsCommitted(*log_.at(index3)));
  EXPECT_FALSE(log_.IsExecutable());
}

TEST_F(LogDeathTest, CommitUntilCase2) {
  auto ballot = 5;
  auto index1 = 1;
  log_.Append(MakeInstance(ballot, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(ballot, index2));
  auto index3 = 3;
  log_.Append(MakeInstance(ballot, index3));

  EXPECT_DEATH(log_.CommitUntil(index3, ballot - 1), "CommitUntil case 2");
}

TEST_F(LogTest, CommitUntilWithGap) {
  auto ballot = 0;
  auto index1 = 1;
  log_.Append(MakeInstance(ballot, index1));
  auto index3 = 3;
  log_.Append(MakeInstance(ballot, index3));
  auto index4 = 4;
  log_.Append(MakeInstance(ballot, index4));

  log_.CommitUntil(index4, ballot);

  EXPECT_TRUE(IsCommitted(*log_.at(index1)));
  EXPECT_FALSE(IsCommitted(*log_.at(index3)));
  EXPECT_FALSE(IsCommitted(*log_.at(index4)));
  EXPECT_TRUE(log_.IsExecutable());
}

TEST_F(LogTest, AppendCommitUntilExecute) {
  std::thread execute_thread([this] {
    log_.Execute();
    log_.Execute();
    log_.Execute();
  });

  auto ballot = 0;
  auto index1 = 1;
  log_.Append(MakeInstance(ballot, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(ballot, index2));
  auto index3 = 3;
  log_.Append(MakeInstance(ballot, index3));

  log_.CommitUntil(index3, ballot);
  execute_thread.join();

  EXPECT_TRUE(IsExecuted(*log_.at(index1)));
  EXPECT_TRUE(IsExecuted(*log_.at(index2)));
  EXPECT_TRUE(IsExecuted(*log_.at(index3)));
  EXPECT_FALSE(log_.IsExecutable());
}

TEST_F(LogTest, AppendCommitUntilExecuteTrimUntil) {
  std::thread execute_thread([this] {
    log_.Execute();
    log_.Execute();
    log_.Execute();
  });

  auto ballot = 0;
  auto index1 = 1;
  log_.Append(MakeInstance(ballot, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(ballot, index2));
  auto index3 = 3;
  log_.Append(MakeInstance(ballot, index3));

  log_.CommitUntil(index3, ballot);
  execute_thread.join();

  log_.TrimUntil(index3);

  EXPECT_EQ(nullptr, log_.at(index1));
  EXPECT_EQ(nullptr, log_.at(index2));
  EXPECT_EQ(nullptr, log_.at(index3));
  EXPECT_EQ(index3, log_.LastExecuted());
  EXPECT_EQ(index3, log_.GlobalLastExecuted());
  EXPECT_FALSE(log_.IsExecutable());
}

TEST_F(LogTest, AppendAtTrimmedIndex) {
  std::thread execute_thread([this] {
    log_.Execute();
    log_.Execute();
  });

  auto ballot = 0;
  auto index1 = 1;
  log_.Append(MakeInstance(ballot, index1));
  auto index2 = 2;
  log_.Append(MakeInstance(ballot, index2));

  log_.CommitUntil(index2, ballot);
  execute_thread.join();

  log_.TrimUntil(index2);

  EXPECT_EQ(nullptr, log_.at(index1));
  EXPECT_EQ(nullptr, log_.at(index2));
  EXPECT_EQ(index2, log_.LastExecuted());
  EXPECT_EQ(index2, log_.GlobalLastExecuted());
  EXPECT_FALSE(log_.IsExecutable());

  log_.Append(MakeInstance(ballot, index1));
  log_.Append(MakeInstance(ballot, index2));
  EXPECT_EQ(nullptr, log_.at(index1));
  EXPECT_EQ(nullptr, log_.at(index2));
}

TEST_F(LogTest, Instances) {
  std::thread execute_thread([this] {
    log_.Execute();
    log_.Execute();
  });

  auto ballot = 0;
  std::vector<multipaxos::Instance> expected;

  expected.emplace_back(MakeInstance(ballot, log_.AdvanceLastIndex()));
  log_.Append(expected.back());
  expected.emplace_back(MakeInstance(ballot, log_.AdvanceLastIndex()));
  log_.Append(expected.back());
  expected.emplace_back(MakeInstance(ballot, log_.AdvanceLastIndex()));
  log_.Append(expected.back());

  EXPECT_EQ(expected, log_.Instances());

  auto index = 2;
  log_.CommitUntil(index, ballot);
  execute_thread.join();
  log_.TrimUntil(index);

  expected.erase(expected.begin(), expected.begin() + index);
  EXPECT_EQ(expected, log_.Instances());
}

TEST_F(LogTest, CallingStopUnblocksExecutor) {
  std::thread execute_thread([this] {
    auto r = log_.Execute();
    EXPECT_EQ(std::nullopt, r);
  });
  std::this_thread::yield();
  log_.Stop();
  execute_thread.join();
}

File: java/replicant/src/test/java/log/LogTest.java
package log;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertNull;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;
import static util.TestUtil.makeInstance;

import command.Command.CommandType;
import command.KVResult;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.Future;
import kvstore.MemKVStore;
import log.Instance.InstanceState;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class LogTest {

  private MemKVStore store_;
  private Log log;


  @BeforeEach
  void SetUp() {
    store_ = new MemKVStore();
    log = new Log(store_);
  }

  @Test
  void constructor() {
    assertEquals(log.getLastExecuted(), 0);
    assertEquals(log.getGlobalLastExecuted(), 0);
    assertFalse(log.isExecutable());
    assertNull(log.at(0L));
    assertNull(log.at(-1L));
    assertNull(log.at(3L));
  }

  @Test
  void insert() {
    HashMap<Long, InstanceWithCv> log = new HashMap<>();
    long index = 1, ballot = 1;
    assertTrue(Log.insert(log, makeInstance(ballot, index, CommandType.Put)));
    assertEquals(CommandType.Put, log.get(index).getInstance().getCommand().getCommandType());
    assertFalse(Log.insert(log, makeInstance(ballot, index, CommandType.Put)));
  }

  @Test
  void insertUpdateInProgress() {
    HashMap<Long, InstanceWithCv> log = new HashMap<>();
    long index = 1, ballot = 1;
    assertTrue(Log.insert(log, makeInstance(ballot, index, CommandType.Put)));
    assertEquals(CommandType.Put, log.get(index).getInstance().getCommand().getCommandType());
    assertFalse(Log.insert(log, makeInstance(ballot, index, CommandType.Put)));
  }

  @Test
  void insertUpdateCommitted() {
    HashMap<Long, InstanceWithCv> log = new HashMap<>();
    long index = 1, ballot = 1;
    assertTrue(
        Log.insert(log, makeInstance(ballot, index, InstanceState.kCommitted, CommandType.Put)));
    assertFalse(
        Log.insert(log, makeInstance(ballot, index, InstanceState.kInProgress, CommandType.Put)));

  }

  @Test
  void insertStale() {
    HashMap<Long, InstanceWithCv> log = new HashMap<>();
    long index = 1, ballot = 1;
    assertTrue(Log.insert(log, makeInstance(ballot, index, CommandType.Put)));
    assertEquals(CommandType.Put, log.get(index).getInstance().getCommand().getCommandType());
    assertFalse(Log.insert(log, makeInstance(0, index, CommandType.Del)));
    assertEquals(CommandType.Put, log.get(index).getInstance().getCommand().getCommandType());
  }

  @Test
  void insertCase2Committed() {
    HashMap<Long, InstanceWithCv> log = new HashMap<>();
    long index = 1;
    var inst1 = makeInstance(0, index, InstanceState.kCommitted, CommandType.Put);
    var inst2 = makeInstance(0, index, InstanceState.kInProgress, CommandType.Del);
    Log.insert(log, inst1);
    var thrown = assertThrows(AssertionError.class, () -> Log.insert(log, inst2));
    assertEquals("Insert case2", thrown.getMessage());
  }

  @Test
  void insertCase2Executed() {
    HashMap<Long, InstanceWithCv> log = new HashMap<>();
    long index = 1;
    var inst1 = makeInstance(0, index, InstanceState.kExecuted, CommandType.Put);
    var inst2 = makeInstance(0, index, InstanceState.kInProgress, CommandType.Del);
    Log.insert(log, inst1);
    var thrown = assertThrows(AssertionError.class, () -> Log.insert(log, inst2));
    assertEquals("Insert case2", thrown.getMessage());
  }

  @Test
  void insertCase3() {
    HashMap<Long, InstanceWithCv> log = new HashMap<>();
    long index = 1;
    var inst1 = makeInstance(0, index, InstanceState.kInProgress, CommandType.Put);
    var inst2 = makeInstance(0, index, InstanceState.kInProgress, CommandType.Del);
    Log.insert(log, inst1);
    var thrown = assertThrows(AssertionError.class, () -> Log.insert(log, inst2));
    assertEquals("Insert case3", thrown.getMessage());
  }

  @Test
  void append() {
    log.append(makeInstance(0, log.advanceLastIndex()));
    log.append(makeInstance(0, log.advanceLastIndex()));
    assertEquals(1, log.at(1L).getIndex());
    assertEquals(2, log.at(2L).getIndex());
  }

  @Test
  void appendWithGap() {
    long index = 42;
    log.append(makeInstance(0, index));
    assertEquals(index, log.at(index).getIndex());
    assertEquals(index + 1, log.advanceLastIndex());
  }

  @Test
  void appendFillGaps() {
    long index = 42;
    log.append(makeInstance(0, index));
    log.append(makeInstance(0, index - 10));
    assertEquals(index + 1, log.advanceLastIndex());
  }

  @Test
  void appendHighBallotOverride() {
    long index = 1, lo_ballot = 0, hi_ballot = 1;
    log.append(makeInstance(lo_ballot, index, CommandType.Put));
    log.append(makeInstance(hi_ballot, index, CommandType.Del));
    assertEquals(CommandType.Del, log.at(index).getCommand().getCommandType());
  }

  @Test
  void appendLowBallotNoEffect() {
    long index = 1, lo_ballot = 0, hi_ballot = 1;
    log.append(makeInstance(hi_ballot, index, CommandType.Put));
    log.append(makeInstance(lo_ballot, index, CommandType.Del));
    assertEquals(CommandType.Put, log.at(index).getCommand().getCommandType());
  }

  @Test
  void commit() {
    long index1 = 1, index2 = 2;
    log.append(makeInstance(0, index1));
    log.append(makeInstance(0, index2));

    assertTrue(log.at(index1).isInProgress());
    assertTrue(log.at(index2).isInProgress());
    assertFalse(log.isExecutable());

    log.commit(index2);

    assertTrue(log.at(index1).isInProgress());
    assertTrue(log.at(index2).isCommitted());
    assertFalse(log.isExecutable());

    log.commit(index1);

    assertTrue(log.at(index1).isCommitted());
    assertTrue(log.at(index2).isCommitted());
    assertTrue(log.isExecutable());
  }

  @Test
  void commitBeforeAppend() {
    long index = 1;
    ExecutorService service = Executors.newFixedThreadPool(1);
    var f = service.submit(() -> log.commit(index));
    Thread.yield();
    log.append(makeInstance(0, log.advanceLastIndex()));
    try {
      assertNull(f.get());
    } catch (InterruptedException | ExecutionException e) {
      throw new RuntimeException(e);
    }
    assertTrue(log.at(index).isCommitted());
  }


  @Test
  void appendCommitExecute() {
    long index = 1;

    ExecutorService service = Executors.newFixedThreadPool(1);

    var f = service.submit(() -> {
      log.execute();
    });

    log.append(makeInstance(0, index));
    log.commit(index);

    try {
      assertNull(f.get());
    } catch (InterruptedException | ExecutionException e) {
      throw new RuntimeException(e);
    }
    assertTrue(log.at(index).isExecuted());
    assertEquals(index, log.getLastExecuted());
  }

  @Test
  void appendCommitExecuteOutOfOrder() {
    ExecutorService service = Executors.newFixedThreadPool(1);
    var f = service.submit(() -> {
      log.execute();
      log.execute();
      log.execute();
    });

    long index1 = 1, index2 = 2, index3 = 3;
    log.append(makeInstance(0, index1));
    log.append(makeInstance(0, index2));
    log.append(makeInstance(0, index3));

    log.commit(index3);
    log.commit(index2);
    log.commit(index1);

    try {
      assertNull(f.get());
    } catch (InterruptedException | ExecutionException e) {
      throw new RuntimeException(e);
    }
    assertTrue(log.at(index1).isExecuted());
    assertTrue(log.at(index2).isExecuted());
    assertTrue(log.at(index3).isExecuted());
    assertEquals(index3, log.getLastExecuted());
  }

  @Test
  void commitUntil() {
    long ballot = 0, index;

    for (index = 1; index < 10; index++) {
      log.append(makeInstance(ballot, index));
    }
    log.append(makeInstance(ballot, index));
    log.commitUntil(index - 1, ballot);

    for (long i = 1; i < index; i++) {
      assertTrue(log.at(i).isCommitted());
    }
    assertFalse(log.at(index).isCommitted());
    assertTrue(log.isExecutable());
  }

  @Test
  void commitUntilHigherBallot() {
    long ballot = 0, index;
    for (index = 1; index < 10; index++) {
      log.append(makeInstance(ballot, index));
    }
    log.commitUntil(index - 1, ballot + 1);

    for (long i = 1; i < index; i++) {
      assertFalse(log.at(i).isCommitted());
    }

    assertFalse(log.isExecutable());
  }

  @Test
  void commitUntilCase2() {
    long ballot = 5, index;
    for (index = 1; index < 10; index++) {
      log.append(makeInstance(ballot, index));
    }

    long finalIndex = index - 1;
    var thrown = assertThrows(AssertionError.class, () -> log.commitUntil(finalIndex, ballot - 1));
    assertEquals("CommitUntil case 2", thrown.getMessage());
  }

  @Test
  void commitUntilWithGap() {
    long ballot = 0, index;
    for (index = 1; index < 10; index++) {
      if (index % 3 == 0) {
        continue;
      }
      log.append(makeInstance(ballot, index));
    }
    log.commitUntil(index - 1, ballot);
    long i;
    for (i = 1; i < index; i++) {
      if (i % 3 == 0) {
        break;
      }
      assertTrue(log.at(i).isCommitted());
    }
    for (; i < index; i++) {
      if (i % 3 == 0) {
        continue;
      }
      assertFalse(log.at(i).isCommitted());
    }
    assertTrue(log.isExecutable());
  }

  @Test
  void appendCommitUntilExecute() {
    ExecutorService executor = Executors.newFixedThreadPool(1);
    List<Future<Map.Entry<Long, KVResult>>> futures = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      futures.add(executor.submit(() -> log.execute()));
    }
    long ballot = 0, index;
    for (index = 1; index < 11; index++) {
      log.append(makeInstance(ballot, index));
    }
    index--;
    log.commitUntil(index, ballot);
    try {
      for (Future<Map.Entry<Long, KVResult>> future : futures) {
        future.get();
      }
    } catch (InterruptedException | ExecutionException e) {
      throw new RuntimeException(e);
    }
    for (long i = 1; i < 11; i++) {
      assertTrue(log.at(i).isExecuted());
    }
    assertFalse(log.isExecutable());
  }

  @Test
  void appendCommitUntilExecuteTrimUntil() {
    ExecutorService executor = Executors.newFixedThreadPool(1);
    List<Future<Map.Entry<Long, KVResult>>> futures = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      futures.add(executor.submit(() -> log.execute()));
    }
    long ballot = 0, index;
    for (index = 1; index < 11; index++) {
      log.append(makeInstance(ballot, index));
    }
    index--;
    log.commitUntil(index, ballot);
    try {
      for (var future : futures) {
        future.get();
      }
    } catch (InterruptedException | ExecutionException e) {
      throw new RuntimeException(e);
    }
    log.trimUntil(index);
    for (long i = 1; i < 11; i++) {
      assertNull(log.at(i));
    }
    assertEquals(index, log.getLastExecuted());
    assertEquals(index, log.getGlobalLastExecuted());
    assertFalse(log.isExecutable());
  }

  @Test
  void appendAtTrimmedIndex() {
    ExecutorService executor = Executors.newFixedThreadPool(1);
    List<Future<Map.Entry<Long, KVResult>>> futures = new ArrayList<>();
    for (int i = 0; i < 10; i++) {
      futures.add(executor.submit(() -> log.execute()));
    }

    long ballot = 0, index;
    for (index = 1; index < 11; index++) {
      log.append(makeInstance(ballot, index));
    }
    index--;
    log.commitUntil(index, ballot);
    try {
      for (var future : futures) {
        future.get();
      }
    } catch (ExecutionException | InterruptedException e) {
      throw new RuntimeException(e);
    }
    log.trimUntil(index);

    for (long i = 1; i < 11; i++) {
      assertNull(log.at(i));
    }
    assertEquals(index, log.getGlobalLastExecuted());
    assertEquals(index, log.getLastExecuted());
    assertFalse(log.isExecutable());

    for (long i = 1; i < 11; i++) {
      log.append(makeInstance(ballot, i));
    }
    for (long i = 1; i < 11; i++) {
      assertNull(log.at(i));
    }
  }

  @Test
  void instances() {
    ExecutorService executor = Executors.newFixedThreadPool(1);
    List<Future<Map.Entry<Long, KVResult>>> futures = new ArrayList<>();
    for (int i = 0; i < 5; i++) {
      futures.add(executor.submit(() -> log.execute()));
    }
    ArrayList<Instance> expected = new ArrayList<>();
    long ballot = 0;
    for (int i = 0; i < 10; i++) {
      expected.add(makeInstance(ballot, log.advanceLastIndex()));
      log.append(expected.get(expected.size() - 1));
    }
    assertEquals(expected, log.instances());

    long index = 5;
    log.commitUntil(index, ballot);
    try {
      for (var future : futures) {
        future.get();
      }
    } catch (ExecutionException | InterruptedException e) {
      throw new RuntimeException(e);
    }

    log.trimUntil(index);
    for (int i = 0; i < index; i++) {
      expected.remove(0);
    }
    assertEquals(expected, log.instances());
  }

  @Test
  void callingStopUnblocksExecutor() {
    ExecutorService executeThread = Executors.newSingleThreadExecutor();
    executeThread.submit(() -> {
      var r = log.execute();
      assertNull(r);
    });
    Thread.yield();
    log.stop();
    executeThread.shutdown();
  }

}